### PR TITLE
test for doctest path problem #8531, and solution

### PR DESCRIPTION
**Pull request by tompscanlan**
* add test reproducing #8531
* re-use solution from #10469
* solution fixes test

### What does this PR try to resolve?

Fixes issue #8531 , wherein we have doctests that fail with a shared library error.

The issue seems painful enough that folks are changing their code to avoid this bug.  We need a solution.

### How should we test and review this PR?

I added a tentative test that reproduces the problem, and reusing #10469, show that solution works.
The repair in question is small. Toggle it off and the test fails with shared library problems.

### CODE REVIEW

File: tests/testsuite/test.rs [hunk 1]
```diff
@@ -48,6 +49,37 @@ fn cargo_test_simple() {
         .run();
 }
 
+#[cargo_test]
+// download a known broken doctests code base and run cargo test, should fail
+fn cargo_test_tensorflow_not_failing_due_to_shared_library() {
+    use git2::build::RepoBuilder;
+
+    // git repo for cargo project containing doctests that require shared libraries
+    let url = "https://github.com/tensorflow/rust";
```
Comment (epage) [inline tests/testsuite/test.rs:58]: Note that we don't rely on network resources in the tests

Comment (tompscanlan) [reply tests/testsuite/test.rs:58]: This was created to reproduce only. I'm hoping you can help me get it shaped into the cargo way.

Comment (tompscanlan) [reply tests/testsuite/test.rs:58]: I'm trying to id a minimal test for this, but it's taking longer than I want, so this was a shortcut to open discussion.

**Comment by tompscanlan** [comment]
I added a test that reproduces the error initially seen in the initial tensorflow issue. I'd like to reduce that to a few files that reproduce the error so I can include it into a unit test.  I'm not having luck, so if anyone wants to take a crack at it, let me know. I can get you details.

I then pulled in @sgrif change from #10469 on top, and agree that the change is making the reproduced error go away. Thanks @sgrif! 

Additionally, I verified that the following can work as a temporary work around if you've got the shared library installed  in your system.  `/usr/loca/lib` is where the library was installed in my system.

    LD_LIBRARY_PATH=/usr/local/lib  cargo test --doc

I feel like adding a build.rs should also be another path to do accomplishing the same thing, but haven't made it work yet.

I'm going to go read #3366 and think on this a little longer.  Next goal is slimming this test down. Anything else you want to see to close this out?

**Comment by tompscanlan** [comment]
I'm spinning my wheels on making a tiny test that replicates the problem.  I removed the test that hits the network. We lose the exact error replication, but the test that remains ensures we have good results assuming you get libs in the right place,.

@epage or @weihanglo , what else should I cover to get this closed?

File: tests/testsuite/test.rs (hunk 1 above)
Comment (tompscanlan) [reply tests/testsuite/test.rs]: removing this. only needed for validation.

**Comment by djc** [comment]
We're running into this for rustls. Seems like a strange bug (potentially a regression?) with potentially a straightforward fix? Cargo team, any hints on moving this forward? I could invest some time to help this along if necessary.

File: src/cargo/core/compiler/compilation.rs [hunk 2]
```diff
@@ -274,7 +274,12 @@ impl<'cfg> Compilation<'cfg> {
         is_rustc_tool: bool,
     ) -> CargoResult<ProcessBuilder> {
         let mut search_path = Vec::new();
+        // this divides our path based on CompileKind: host vs target.
         if is_rustc_tool {
+            search_path.extend(super::filter_dynamic_search_path(
```
Comment (weihanglo) [inline src/cargo/core/compiler/compilation.rs:279]: I believe you've read this comment <https://github.com/rust-lang/cargo/pull/10469#issuecomment-1079503079> and [this zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/dynamic.20linkers.20.26.20mixed.20targets). 

If I understand correctly, the issue is like

1. `rustdoc --test` does two things: compiling the example and running it
2. a proc-macro dep uses a shared lib “foo.so” from system
3. A build script builds a shared lib “foo.so” from vendored sources
4. Name collision happens as proc-macro is supposed to use "foo.so" from system not vendored one.

This should be unlikely as ehuss stated out:

> In general, using shared libraries with proc-macros has some bugs, so hopefully people don't actually use them too often. For example, I just discovered that `cargo doc` does not set the search path for build script OUT directories, so any proc-macro that tries to load a shared library generated by a build script will fail.

That being said, we can propably have patch like this PR and document the truth that rustc/cargo don't support proc-macro loading shared libs at this moment.

**Review by weihanglo — COMMENTED** [review]
Sorry for the delay. While I think a proper solution would rustdoc delegate doctests runnning to cargo (which should be tracked in <https://github.com/rust-lang/testing-devex-team/issues/5>), we should try moving this forwards, or at least have a conclusion what to do next. Will put this in the agenda of the triage meeting.

File: src/cargo/core/compiler/compilation.rs [hunk 3]
```diff
@@ -274,7 +274,12 @@ impl<'cfg> Compilation<'cfg> {
         is_rustc_tool: bool,
     ) -> CargoResult<ProcessBuilder> {
         let mut search_path = Vec::new();
+        // this divides our path based on CompileKind: host vs target.
```
[the remark this answers is not in the document]
Comment (tompscanlan) [reply src/cargo/core/compiler/compilation.rs]: comment updated

File: src/cargo/core/compiler/compilation.rs (hunk 2 above)
Comment (tompscanlan) [reply src/cargo/core/compiler/compilation.rs:279]: I'll defer to the team. I'm just carrying the change forward from #10469.

**Pull request closed without merging**